### PR TITLE
Add pre-commit hooks

### PR DESCRIPTION
**Pull request by fsimonis**
## Main changes of this PR

This PR adds pre-commit hooks to preCICE including:

* formatting of code using the correct version of clang-format
* formatting of [preCICE config files](https://github.com/precice/precice-pre-commit-hooks)
* checks for merge conflict markers
* checks for mixed line-endings in files
* checks for incorrect XML files (ignores namespaces :wink:)

This PR replaces the code check workflow and doesn't rely on https://pre-commit.ci/

## Motivation and additional information

Massive shoutout to @renefritze for showing us this tool!

Closes #1291

## Author's checklist

* [ ] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+repos:
```
Comment (MakisH) [inline .pre-commit-config.yaml:1]: Where's the docs? :grin: 

To document/steps:
- This is a convenience tool, not a safety measure: you need to explicitly install pre-commit if you want Git to warn you before you commit.
- Install the tool: `pip3 install pre-commit`
- Install the hook: `pre-commit install`

When committing something good, it shows the following output:
```
~/github/precice [add-pre-commit]$ git commit
clang-format.............................................................Passed
check xml............................................(no files to check)Skipped
check for merge conflicts................................................Passed
mixed line ending........................................................Passed
format preCICE configuration files...................(no files to check)Skipped
[add-pre-commit f886de24] test
 1 file changed, 1 insertion(+), 1 deletion(-)
```

**Review by MakisH — APPROVED** [review]
I tried it locally and it works. I also understand the code changes. I don't think we need to add any other checks at the moment.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (fsimonis) [reply .pre-commit-config.yaml:1]: It's not only a convenience tool, you can use it to format your files using the correct version of clang-format, which is super useful if your distro doesn't provide it.

Comment (renefritze) [reply .pre-commit-config.yaml:1]: Just FYI, this is thanks to a binary wheel distribution of clang-format via https://github.com/ssciwr/clang-format-wheel

**Comment by fsimonis** [comment]
I extended the formatting readme. Problem is that there is no place for this doc in the source-itself.
I'll add it to the developer documentation of the website.

**Pull request merged**